### ISSUE TITLE
Read-Only API

### ISSUE THREAD

**Issue by nbauma109** [ib-controller/ib-controller#31]
Hello,

In the newest version of the Gateway, there is a new option Read-Only API.
As it was unexpectedly checked last time I used the Gateway and as orders wouldn't pass for that reason, I'd like a new property to set this setting.
I think it should be false by default.
Could this be achieved in this task : ConfigureTwsApiPortTask ?

Kind regards, 
Nicolas.

**Comment by rlktradewright** [comment]
Nicolas

I'm not convinced that there is a need for this. I understand that the new setting caught you out somehow, but now that's happened to you I doubt you'll ever be caught again. 

In my view this is one of those settings that you set manually when you first configure the API, and after that you never touch it. If you need to sometimes run with it set differently, the best way is to have a separate IBController configuration for this (ie a separate start script, and a separate .ini file with a different TWS settings directory).

I'm not saying definitely no, just that you'll need to convince me that the effort to code, test, document  and forever maintain this setting is worth it.

**Comment by nbauma109** [comment]
Hi Richard, 

In fact I believe that most of the settings are set once and for all in the properties file that ensures that you are in control of them. At least this is how I have always been using ib controller. 
If you're not convinced, I may develop this for myself. 

Kind regards, 
Nicolas.

**Comment by rlktradewright** [comment]
I have to disagree. Hardly any of the TWS API settings are controllable via IBController. It would of course be technically feasible to use IBController to set all of them, but this is simply pointless. What benefit would it give? And the more of this stuff is in IBController, the more there is to maintain and test with each new release. Not to mention that these things are very subject to being broken by IB changing the API configuration dialog (or the menu path to it) which has happened in the past.

If you are setting up a system for live trading, then you should treat the system the same as any production system: set it up, test it, then don't mess with it unless it's absolutely necessary. When you do change anything, you MUST test again before running it live (particularly if you start to use a new version of TWS!).

So you should never have any surprises with your live system.

For paper trading systems, you can be a bit more relaxed, because you're not going to lose money if it doesn't work as expected. And there are many ways you can change settings in TWS to stop it working the way you want.

The bottom line is that IBController is not intended to be an insurance policy against configuration mistakes in TWS. It's purpose is to automate things that need to be automated because there is no other way to achieve the desired end.

By all means implement this if you want to. If you do, you can even submit a pull request and I'll give it due consideration. 

One suggestion if you do decide to implement it: I would make this a three-valued option:

* read only

* read-write

* whatever is currently set in TWS (which should be the default)

**Comment by rlktradewright** [comment]
In view of no further comment, I'll close this.

**Comment by Fynjy** [comment]
Sorry for offtop, but my question relates to this issue:
I turned off the read-only mode via Gateway's GUI (edemo/demouser account), but I have a problem with restore of settings. Error message:
20:43:28:531 IBController: TWS settings directory is /home/user/api/settings
...
20:58:37:880 AWT-EventQueue-0: Settings : loadFromXMLFile : Could not find settings file /home/user/api/settings/darykqwzu/ibg.xml
IbDir=/home/user/api/settings, but new subdir like a 'darykq*' is created for the each session. This subdir is used for load/store settings.

The same if I run Gateway by command 'java -cp jts.jar:total.2013.jar -Dsun.java2d.noddraw=true -Xmx512M ibgateway.GWClient'.

**Comment by rlktradewright** [comment]
I'm sorry to hear that you have a problem running the gateway. However since this problem seems to have nothing to do with IBController, I think you're asking in the wrong place. 

Having said that, you could try just deleting the gateway settings directory (or rename it so you can restore it again if need be) and starting with a blank slate. I had to do this the first time I ran the latest version of TWS.

**Comment by nbauma109** [comment]
I just got caught again by the setting! It went to back to read only for an unknown reason :-/ 
I did nothing special... 
I guess I'll have no choice but to handle it.

**Comment by rlktradewright** [comment]
I haven't had a problem yet with Read-only API, but I did have similar problems with the 'allow connections only from localhost' setting. I found that sometimes, if 'allow connections only from localhost' was not set and I added a new entry to the trusted IP address list, TWS would then not allow connections from anywhere, and I'd have to stop TWS, wipe out TWS's settings folder, and start all over again. I eventually managed to get all the addresses I need (about 7 of them) configured and working. Unfortunately I couldn't identify the circumstances that made this happen - when I tried to reproduce it so I could report the bug to IB, I couldn't make it fail again! So I haven't reported it yet.

I certainly won't use this version in my live systems till I'm sure it's not going to bite me. I'll have to try experimenting again so I can give IB some solid evidence.

**Comment by feci** [comment]
Hi,

I have the same problem... Setting is on by default.
I use ib-controller in cloud server environment (linux with Xvfb - instead of desktop environment).
New servers with tws are started often, while no settings is shared. There is no manual intervention possible, as there is no desktop environment.

I agree that this setting should be switched to off by default. Possibly by ib-controller?

**Comment by rlktradewright** [comment]
@feci I don't understand how you manage to start up new servers containing TWS with no configuration. There are so many settings to do with the API: are you saying that you are completely happy with all the defaults, except Read-Only API? 

How do you even install TWS on these servers? How do you edit the IBController.ini file to include the IB username and password? 

I presume the answer is that you do all these things via an RSH session or similar. But in that case, why don't you just configure a IBController + TWS setup locally, and simply copy all the files over to the target system?

I really do NOT want IBController to turn into a TWS configuration mechanism without a genuine justification.

Regarding the IB silliness of making Read-Only API true by default, the correct thing to do is to complain to IB about it and get them to change the default. There are a great many TWS users who don't use IBController, and I suspect they're all just as annoyed about this as we are, so let's get on to IB about it.

**Comment by rlktradewright** [comment]
In view of lack of further discussion of this issue, I'm closing it.
